cmd/kube-controller-manager/app/cloudproviders.go

```diff
@@ -32,29 +30,28 @@ import (
 func createCloudProvider(logger klog.Logger, cloudProvider string, externalCloudVolumePlugin string, cloudConfigFile string,
 	allowUntaggedCloud bool, sharedInformers informers.SharedInformerFactory) (cloudprovider.Interface, ControllerLoopMode, error) {
 	var cloud cloudprovider.Interface
-	var loopMode ControllerLoopMode
 	var err error
-
-	if utilfeature.DefaultFeatureGate.Enabled(features.DisableCloudProviders) && !cloudprovider.IsExternal(cloudProvider) {
-		cloudprovider.DisableWarningForProvider(cloudProvider)
-		return nil, ExternalLoops, fmt.Errorf(
-			"cloud provider %q was specified, but built-in cloud providers are disabled. Please set --cloud-provider=external and migrate to an external cloud provider",
-			cloudProvider)
-	}
+	loopMode := ExternalLoops
 
 	if cloudprovider.IsExternal(cloudProvider) {
-		loopMode = ExternalLoops
 		if externalCloudVolumePlugin == "" {
 			// externalCloudVolumePlugin is temporary until we split all cloud providers out.
 			// So we just tell the caller that we need to run ExternalLoops without any cloud provider.
 			return nil, loopMode, nil
 		}
 		cloud, err = cloudprovider.InitCloudProvider(externalCloudVolumePlugin, cloudConfigFile)
 	} else {
-		cloudprovider.DeprecationWarningForProvider(cloudProvider)
-
-		loopMode = IncludeCloudLoops
-		cloud, err = cloudprovider.InitCloudProvider(cloudProvider, cloudConfigFile)
+		// in the case where the cloudProvider is not set, we need to inform the caller that there
+		// is no cloud provider and that the default loops should be used, and there is no error.
+		// this will cause the kube-controller-manager to start the default controller contexts
+		// without being attached to a specific cloud.
+		if len(cloudProvider) == 0 {
+			loopMode = IncludeCloudLoops
```

Comment:
I have doubts we should run this controllers https://github.com/kubernetes/kubernetes/pull/127711/commits/239105b37f47c600c0d6e069573b86e1a5d86ee2#r1780627847 in any case, all of them require a cloud provider config

Reply:
i agree and i would like to follow up with more cleanups, but this seemed like a reasonable start.

Reply:
@elmiko we need to remove this, the cloud provider is initialized in the line above`		cloud, err = cloudprovider.InitCloudProvider(externalCloudVolumePlugin, cloudConfigFile)` and this is enabling the cloud provider controllers with a nil cloud provider

CloudProvider name can not be empty, so this seems to enable the InnerLoops if there is no cloud provider set (no cloud provider flag) (we need to add tests in cmd/kube-controller-manager/app/options/options_test.go)

Reply:
I found the flag 

https://github.com/kubernetes/kubernetes/blob/ec16c90aaf5d1f0606747c421c8680bb2b243d4e/staging/src/k8s.io/cloud-provider/options/cloudprovider.go#L36-L38

and it says

> 		"The provider for cloud services. Empty string for no provider.")


so if the name is empty is no provider, hence no controllers :)

I will follow up

Reply:
thanks @aojea , i agree it should be removed. i had left it in place to preserve some of the original flow, but yes the way `cloud, err` is set, this shouldn't be needed. i think i might have seen some odd behavior when i removed it, but i don't recall. regardless, happy for more cleanups ;)